pkg/store/storepb/shard_info.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package storepb
+
+import (
+	"github.com/alecthomas/units"
+	"github.com/cespare/xxhash/v2"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+)
+
+var sep = []byte{'\xff'}
+
+type ShardMatcher struct {
+	buf              []byte
+	shardingLabelset map[string]struct{}
+
+	isSharded bool
+
+	by          bool
+	totalShards int64
+	shardIndex  int64
+}
+
+func (s *ShardMatcher) IsSharded() bool {
+	return s.isSharded
+}
+
+func (s *ShardMatcher) MatchesZLabels(zLabels []labelpb.ZLabel) bool {
+	// Match all series when query is not sharded
+	if s == nil || !s.isSharded {
+		return true
+	}
+
+	s.buf = s.buf[:0]
+	for _, lbl := range zLabels {
+		// Exclude metric name and le label from sharding
+		if lbl.Name == "__name__" || lbl.Name == "le" {
+			continue
+		}
+
+		if shardByLabel(s.shardingLabelset, lbl, s.by) {
+			s.buf = append(s.buf, lbl.Name...)
+			s.buf = append(s.buf, sep[0])
+			s.buf = append(s.buf, lbl.Value...)
+			s.buf = append(s.buf, sep[0])
+		}
+	}
+
+	hash := xxhash.Sum64(s.buf)
+	return hash%uint64(s.totalShards) == uint64(s.shardIndex)
+}
+
+func (s *ShardMatcher) MatchesLabels(lbls labels.Labels) bool {
+	return s.MatchesZLabels(labelpb.ZLabelsFromPromLabels(lbls))
+}
+
+func shardByLabel(labelSet map[string]struct{}, zlabel labelpb.ZLabel, groupingBy bool) bool {
+	_, shardHasLabel := labelSet[zlabel.Name]
+	if groupingBy && shardHasLabel {
+		return true
+	}
+
+	groupingWithout := !groupingBy
+	if groupingWithout && !shardHasLabel {
+		return true
+	}
+
+	return false
+}
+
+func (m *ShardInfo) Matcher() *ShardMatcher {
+	if m == nil || m.TotalShards < 1 {
+		return &ShardMatcher{
+			isSharded: false,
+		}
+	}
+
+	return &ShardMatcher{
+		isSharded:        true,
+		buf:              make([]byte, 10*units.Kilobyte),
```

Comment:
I think we should introduce a private `sync.Pool` for this and make a function Close() that returns this buffer to the pool. I think everywhere we have code like:
```go
	shardMatcher := r.ShardInfo.Matcher()
	...
	do something with shardMatcher
```

So, as far as I can tell it should be trivial to use a sync.Pool here:

```go
	shardMatcher := r.ShardInfo.Matcher()
	defer shardMatcher.Close()
```

Reply:
Sounds good, will do that.

Reply:
I made this change and pushed a new commit. Could you take a look if this is what you had in mind?